Fix/refactor search by actor

Noor-Serry opened the pull request:
# Pull Request Template

## Description
Provide a clear and concise description of what this pull request does. Include the purpose and context for the changes.
refactor the search by actor view model and his state and error handling make the test coverage 100 %
---

## Changes Made
List the changes introduced in this pull request:

- Remove NoInternetConnection from SearchByActorEffect since it’s not an effect
- Rename onRetryRequestClicked to a clearer name following our naming conventions
- Rename onUserSearch interaction function to a clearer name like onChangeSearchText   
- fix the test
---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.
<img width="752" height="61" alt="image" src="https://github.com/user-attachments/assets/4f93b834-7691-45ff-8659-c3d0d62c54b2" />


---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

marah222 committed:
ref : moved  and renamed search by actor package under Search package

marah222 committed:
ref : renaming any query as keyword

marah222 committed:
ref : renaming any query as keyword in ViewModel & Screen State

marah222 committed:
ref : extract searchMoviesByActorError function

marah222 committed:
Merge branch 'refs/heads/develop' into fix/refactor-search-by-actor-screen-UI-viewModel-state

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt

marah222 committed:
ref : renamed Action listener

marah222 committed:
ref : replaced not internet effect with no internet state

marah222 committed:
ref : search by Actor viewModel Test

marah222 committed:
ref : formatting

eslam-99 committed:
fix: fix states to be centered in screen in SearchByActorScreen.kt

marah222 committed:
Merge branch 'develop' into fix/refactor-search-by-actor

marah222 committed:
fix : refactor ui state handling

eslam-99 committed:
Merge remote-tracking branch 'origin/develop' into fix/refactor-search-by-actor

Commit:
refactor: improve SearchByActorScreen and ViewModel

This commit refactors the `SearchByActorScreen` and its corresponding `SearchByActorViewModel`.

Key changes include:
- Renamed `keyword` to `query` in `SearchByActorScreenState` and `SearchByActorViewModel`.
- Updated interaction listener methods and their implementations.
- Improved error handling and UI states for network errors and no results.
- Added navigation to movie details screen when a movie is clicked.
- Removed `ActorSearchErrorState.kt` as its functionality is now handled within the ViewModel.
- Updated UI layout and components in `SearchByActorScreen.kt`.

Commit:
Merge branch 'feature/movie_serise_details' of github.com:Amsterdam-Team/Aflami into fix/refactor-search-by-actor

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt
#	viewModel/src/main/java/com/example/viewmodel/search/actorSearch/SearchByActorScreenState.kt
#	viewModel/src/main/java/com/example/viewmodel/searchByActor/SearchByActorEffect.kt
#	viewModel/src/main/java/com/example/viewmodel/searchByActor/SearchByActorInteractionListiner.kt
#	viewModel/src/test/java/com/example/viewmodel/SearchByActorViewModelTest.kt
#	viewModel/src/test/java/com/example/viewmodel/utils/entityHelper/createMovie.kt

Commit:
refactor: relocate searchByActor files

Moved SearchByActorEffect.kt and SearchByActorInteractionListener.kt to the `search.actorSearch` package.

Updated imports in SearchByActorScreen.kt to reflect the new location.
Simplified `SearchByActorInteractionListener` by removing `onNavigateBackClicked`, `onRetryQuestClicked`, and `onMovieClicked` methods and adding `onNavigateBackClick`.

Commit:
refactor: rename methods in SearchByActorViewModel

This commit refactors `SearchByActorViewModel` by renaming several methods for clarity and consistency.

- `onKeywordValueChanged` is now `onUserSearchChange`.
- `onNavigateBackClicked` is now `onNavigateBackClick`.
- `onRetryQuestClicked` is now `onRetrySearchClick`.

These changes are also reflected in the `SearchByActorInteractionListener` interface and updated in the `SearchByActorViewModelTest`.

Commit:
refactor: remove NoCastFoundException and its UI state

This commit removes the `NoCastFoundException` and its corresponding UI state `CastErrorUiState.NoCastFound`.

The logic for handling cases where no cast is found has been updated in `CastScreen.kt`. Instead of relying on a specific error state, it now checks if `state.cast` is empty to display the "cast not available" message.

Additionally, the `getActorsByMovieId` function in `MovieRepositoryImpl.kt` has been simplified.

Commit:
refactor: rename `query` to `keyword` in SearchByActor

This commit renames the `query` property to `keyword` in the
`SearchByActorScreenState` and related components. It also updates
the corresponding interaction listener method from `onUserSearch` to
`onUserSearchChange` for better clarity.

The tests have been updated to reflect these changes and ensure
continued functionality. Additionally, new tests have been added to cover
the behavior of `onMovieClicked` and handling of unknown errors during
search.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into fix/refactor-search-by-actor

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/cast/CastScreen.kt
#	viewModel/src/main/java/com/example/viewmodel/cast/CastUiState.kt

Commit:
refactor: improve error handling in SearchByActor screen

Refactored the error handling in `SearchByActorScreen` and `SearchByActorViewModel`.

- Replaced the `isNetworkError` boolean flag with an `error` property of type `SearchByActorError?` in `SearchByActorScreenState`.
- `SearchByActorError` is a new sealed class to represent different error types.
- Updated `SearchByActorViewModel` to set the new `error` property based on the type of exception.
- Updated `SearchByActorScreen` to display the appropriate error UI based on the `error` state.
- Adjusted unit tests to reflect these changes.

Additionally, renamed `movieLength` to `runTime` in the `createMovie` test helper function for consistency.

Belal-Alkhatib commented on domain/src/main/java/com/example/domain/useCase/GetMoviesByActorUseCase.kt:
```diff
@@ -10,5 +11,8 @@ class GetMoviesByActorUseCase(
     suspend operator fun invoke(actorName: String): List<Movie> {
         return movieRepository.getMoviesByActor(actorName)
             .sortByPopularityDescending()
+//            .ifEmpty {
+//                throw NoSearchByActorResultFoundException()
+//            }
```
dead code

Belal-Alkhatib commented on viewModel/src/main/java/com/example/viewmodel/search/actorSearch/SearchByActorViewModel.kt:
```diff
@@ -27,66 +29,74 @@ class SearchByActorViewModel(
+    private fun observeActorSearchQuery() {
         viewModelScope.launch {
-            queryFlow
+            keywordFlow
```
we already have an extension function that do the same work, i already refactor it in ui-viewmodel-module branch

Belal-Alkhatib commented on viewModel/src/main/java/com/example/viewmodel/search/actorSearch/SearchByActorScreenState.kt:
```diff
@@ -0,0 +1,16 @@
+
+import com.example.viewmodel.search.countrySearch.MovieUiState
+
+data class SearchByActorScreenState(
```
this class should be renamed to ActorSearchUiState to stay consistent with the rest of our UI state classes, note that i already refactor it in ui-viewmodel-module branch

Belal-Alkhatib reviewed (changes requested):
I’ve added some refactor changes in the ui-viewmodel-module branch. Please review them to avoid potential merge conflicts.

saraashraf11 commented on ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt:
```diff
@@ -98,12 +83,12 @@ private fun SearchByActorContent(
             title = stringResource(R.string.find_by_actor),
             showNavigateBackButton = true,
-            onNavigateBackClicked = { interactionListener.onNavigateBackClicked() }
+            onNavigateBackClicked = interactionListener::onNavigateBackClick
```
I suggest changing this name to onClicked at first as Barq said.

saraashraf11 commented on ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt:
```diff
@@ -187,16 +177,18 @@ private fun SearchByActorContentPreview() {
-                override fun onNavigateBackClicked() {
+
+                override fun onNavigateBackClick() {
                 }
```
Change the name on all of this.

Noor-Serry replied:
<img width="767" height="475" alt="image" src="https://github.com/user-attachments/assets/87fdd095-9573-4d79-bc70-a4bc092362d5" />

Noor-Serry replied:
it should be I changed it .

saraashraf11 commented on viewModel/src/test/java/com/example/viewmodel/SearchByActorViewModelTest.kt:
```diff
@@ -1,94 +1,214 @@
+        Dispatchers.resetMain()
+    }
+
+    @Test
```
The test message should return... when...
For the sake of unity among us all

Noor-Serry replied:
that its wrong that way will work will if we have only one public function like in the usecases but here where are have different function so I need to write the name of the function first .

saraashraf11 commented on viewModel/src/test/java/com/example/viewmodel/SearchByActorViewModelTest.kt:
```diff
@@ -1,94 +1,214 @@
+    }
+
+    @Test
+    fun `onUserSearchChange should update keyword and set loading to false when onKeywordValueChanged is called with blank string`() =
```
It is preferable to replace = with brackets for the sake of consistency in writing the code between us.

Noor-Serry replied:
I Checked the other view model test and it follow the same way that I have did

saraashraf11 commented:
resolve conflict please

Commit:
refactor: Standardize naming in SearchByActor feature

This commit refactors the naming conventions within the SearchByActor feature for consistency and clarity.

Specifically, it renames:
- `SearchByActorViewModel` to `SearchActorViewModel`
- `SearchByActorScreenState` to `ActorSearchUiState`
- `SearchByActorEffect` to `SearchActorEffect`
- `SearchByActorInteractionListener` to `SearchActorInteractionListener`

Additionally, method names within the interaction listener have been updated to use the "onClick" prefix (e.g., `onNavigateBackClick` to `onClickNavigateBack`).

The `debounceSearch` utility function is now used for debouncing search queries.

Finally, the `ifEmpty` block within the `GetMoviesByActorUseCase` has been removed.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into fix/refactor-search-by-actor

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt

Commit:
fix: correct movie card type

The movie type displayed in the `MovieCard` in `SearchByActorScreen` was previously hardcoded as "Movies". This commit changes it to use the localized string resource `R.string.movie`.

Additionally, the `MovieImage` composable was introduced to handle image loading using `SafeImageView`.

Commit:
refactor: pass movie ID directly in NavigateToDetailsScreen effect

The `NavigateToDetailsScreen` effect in `SearchActorEffect` now directly carries the `movieId`.

This change simplifies the `SearchActorViewModel` by removing the need to store the `selectedMovieId` in the `ActorSearchUiState`. Consequently, the `ActorSearchUiState` no longer includes the `selectedMovieId` field.

The `SearchByActorScreen` has been updated to retrieve the `movieId` from the effect itself when navigating.

Unit tests for `SearchActorViewModel` have been adjusted to reflect this change in how the `movieId` is passed.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into fix/refactor-search-by-actor

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	ui/src/main/java/com/example/ui/screens/search/actorSearch/SearchByActorScreen.kt
#	viewModel/src/main/java/com/example/viewmodel/search/actorSearch/ActorSearchInteractionListener.kt
#	viewModel/src/main/java/com/example/viewmodel/search/actorSearch/ActorSearchUiState.kt
#	viewModel/src/main/java/com/example/viewmodel/search/actorSearch/ActorSearchViewModel.kt
#	viewModel/src/test/java/com/example/viewmodel/SearchByActorViewModelTest.kt

Commit:
Refactor: Relocate BaseViewModel and update SearchActorViewModel

This commit refactors the `BaseViewModel` by moving it to the `shared` package.
Consequently, `SearchActorViewModel` has been updated to import `BaseViewModel` from its new location.

Additionally, the `toListOfUiState` mapper function in `SearchActorViewModel` has been renamed to `toMoveUiStates` for clarity.
The unused `MediaItemUiState.kt` file has been removed.

Commit:
refactor: rename ViewModel and update UI state for actor search

This commit refactors the actor search functionality by:

- Renaming `SearchByActorViewModel` to `SearchActorViewModel` in the dependency injection module.
- Replacing `MovieUiState` with `MovieItemUiState` in `ActorSearchUiState`.
- Updating `SearchByActorScreen` to use the properties of `MovieItemUiState` (e.g., `posterImageUrl`, `yearOfRelease`, `rate`) for displaying movie information.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into fix/refactor-search-by-actor

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/search/actorSearch/SearchByActorScreen.kt

Commit:
add loading and error indicators to actor search images

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/search/actorSearch/SearchByActorScreen.kt:
```diff
@@ -17,76 +17,68 @@ import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
+                    navController.navigate(Route.MovieDetails(effect.movieId))
+                }
+
+                null -> {}
```
you can use `let{}` check to avoid null

Commit:
feat: migrate SearchByActorScreen to Orbit MVI

This commit migrates the `SearchByActorScreen` to use Orbit MVI.
The `LaunchedEffect` now collects `viewModel.effect` and handles navigation effects.
The UI state is observed and updated accordingly using `Crossfade`.
Loading, empty, and success states are handled with appropriate UI elements.
The `LazyVerticalGrid` for displaying movies is updated with `contentPadding`.

nabilBouzineDev commented on ui/src/main/java/com/example/ui/screens/search/actorSearch/SearchByActorScreen.kt:
```diff
@@ -121,19 +113,26 @@ private fun SearchByActorContent(
+                                        .align(Alignment.CenterHorizontally)
+                            )
+
+                        null -> {}
```
Please avoid adding null here

Noor-Serry replied:
oh that is best Comment I have received . good job

Commit:
fix: handle null error case in SearchByActorScreen

The `when` statement for handling errors in `SearchByActorScreen`
was missing a case for `null` errors. This change adds an `if`
statement to specifically handle `NetworkError` and implicitly ignores
other error types, including `null`, thus preventing a potential crash.
